### PR TITLE
HPCC-30457 ParquetIO.Read() improperly combines hThor and Thor files

### PR DESCRIPTION
**Pull request by jackdelv**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: plugins/parquet/parquetembed.cpp [hunk 1]
```diff
@@ -227,13 +227,48 @@ arrow::Status ParquetHelper::openReadFile()
         StringBuffer filename;
         StringBuffer path;
         splitFilename(location.c_str(), nullptr, &path, &filename, nullptr, false);
-        Owned<IDirectoryIterator> itr = createDirectoryIterator(path.str(), filename.append("*.parquet"));
+        Owned<IDirectoryIterator> fileItr;
+
+        // hThor and Thor files that were written with parquet have diffent naming schemes
```
Comment (ghalliday) [inline plugins/parquet/parquetembed.cpp:232]: typo: different

File: plugins/parquet/parquetembed.cpp [hunk 2]
```diff
@@ -227,13 +227,48 @@ arrow::Status ParquetHelper::openReadFile()
...
+        // Thor files are tagged with a worker ID while hThor files are single non-tagged files
+        if (activityCtx->numSlaves() > 1)
+        {
+            // Executing on Thor: Check for single files that were written by hThor or from external source
+            Owned<IDirectoryIterator> singleItr = createDirectoryIterator(path.str(), filename.append(".parquet"));
+            if (!singleItr || !singleItr->first())
```
Comment (ghalliday) [inline plugins/parquet/parquetembed.cpp:238]: minor: createDirectoryIterator() never returns null, so no need to check.

File: plugins/parquet/parquetembed.cpp [hunk 3]
```diff
@@ -227,13 +227,48 @@ arrow::Status ParquetHelper::openReadFile()
...
+            {
+                // Check for partitioned files which will have indexes prefixed to the filename
+                Owned<IDirectoryIterator> multItr = createDirectoryIterator(path.str(), filename.insert(filename.length() - 8, '*'));
+                fileItr = multItr;
+            }
+            else
+            {
+                // There was a single file so check for partitioned files
+                Owned<IDirectoryIterator> multItr = createDirectoryIterator(path.str(), filename.insert(filename.length() - 8, "[0-9]*"));
```
Comment (ghalliday) [inline plugins/parquet/parquetembed.cpp:247]: I don't think this does what you expect.  "[0-9]\*" is not treated as a regular expression it is a filename wildcard containing '\*' or '?'.  I think you will need to check for filename* and then post-process.  Or use "filename*.parquet"?

File: plugins/parquet/parquetembed.cpp [hunk 4]
```diff
@@ -227,13 +227,48 @@ arrow::Status ParquetHelper::openReadFile()
...
+                if (!multItr || !multItr->isValid())
+                    fileItr = singleItr; // If there aren't any partitioned files read the single file with Thor
```
Comment (ghalliday) [inline plugins/parquet/parquetembed.cpp:249]: why not unconditionally assign multiItr, neither of them contain any values

**Review by ghalliday — CHANGES_REQUESTED** [review]
@dcamper 
@jackdelv I'm not sure this is the right solution.
A more consistent solution for the way other file operations work would be to check for any "filename*.parquet" files on output.  If they exist, then fail to write - unless OVERWRITE is specified, in which case they should be deleted.

**Comment by jackdelv** [comment]
@ghalliday I am closing this pull request to address this issue in HPCC-30456 as it is closely related.

**Pull request closed without merging**